flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1277,6 +1293,34 @@ private void discardCheckpoint(PendingCheckpoint pendingCheckpoint, @Nullable Th
 		}
 	}
 
+	private void tryHandleCheckpointException(DeclineCheckpoint message, PendingCheckpoint currentCheckpoint) {
+		boolean needFailJob = message.getReason() != null && !(message.getReason() instanceof CheckpointDeclineException)
+			&& failOnCheckpointingErrors && !currentCheckpoint.getProps().isSavepoint();
+
+		if (needFailJob) {
+			boolean processed = false;
+			ExecutionAttemptID failedExecutionID = message.getTaskExecutionId();
+			for (ExecutionVertex vertex : tasksToWaitFor) {
+				if (vertex.getCurrentExecutionAttempt().getAttemptId().equals(failedExecutionID)) {
+					if (currentPeriodicTrigger != null) {
+						currentPeriodicTrigger.cancel(true);
+						currentPeriodicTrigger = null;
+					}
+
+					vertex.fail(message.getReason());
```

Comment:
@aljoscha Can you help determine if this operation will have side effects? cc @azagrebin

Reply:
We can't make `Execution` or `ExecutionGraph` fail in `CheckpointCoordinator`, so we can throw out the decision.